Buchsbaum frequency

svincena opened the pull request:
A Hack Week version of the code from the "frequency" breakout room that attempts to add the buchsbaum_frequency function for calculating said frequency in a two-ion species, multi-fluid, magnetized plasma. The frequency occurs at a point when the perpendicular dielectric tensor element goes to zero due to the 180 degree out of phase motion for the two ion fluids.

svincena committed:
Create ide_spyder.md

Initial, bare-bones spyder ide resource

svincena committed:
Merge branch 'PlasmaPy:main' into main

svincena committed:
Add initial Alfven_speed function via Alfven_speed.py

svincena committed:
Add initial thoughts about a Buchsbaum frequency for plasmas with two ion species and a neutralizing electron fluid.

svincena committed:
Merge branch 'PlasmaPy:main' into Buchsbaum_frequency

svincena committed:
Merge remote-tracking branch 'origin/Buchsbaum_frequency' into Buchsbaum_frequency

svincena committed:
Delete Alfven_speed.py

svincena committed:
Add initial thoughts about a Buchsbaum frequency for plasmas with two ion species and a neutralizing electron fluid.

svincena committed:
Adding more details to Buchsbaum frequency code. Didn't have the density as an argument for example.

svincena committed:
Code touchups

svincena committed:
Merge branch 'PlasmaPy:main' into main

svincena committed:
Adding decorators: angular_freq_to_hz and @validate_quantities. Minor docstring cleanups.

svincena committed:
Changed squared def of Buchsbaum_frequency for sqrt on RHS to avoid negative root ambiguity

svincena committed:
Merge branch 'main' into Buchsbaum_frequency

svincena committed:
Put back Alfven speed

namurphy commented on hack/Buchsbaum_frequency.py:
```diff
@@ -36,35 +47,50 @@ def buchsbaum_frequency(B: Unit('T'), n1: u.m**-3, n2: u.m**-3, particle1: Parti
+        If units are not provided, SI units are assumed.
         Notes
         -----
 
```
```suggestion
        If units are not provided, SI units are assumed.

    Notes
    -----
```

namurphy commented on hack/Buchsbaum_frequency.py:
```diff
@@ -36,35 +47,50 @@ def buchsbaum_frequency(B: Unit('T'), n1: u.m**-3, n2: u.m**-3, particle1: Parti
     component of the cold-plasma dielectric coefficient
     math:: \epsilon_{\perp} to vanish at an angular frequency referred to as the
-    Buchsbaum frequency[1], also called the bi-ion frequency or ion-ion hybrid frequency[2].
+    Buchsbaum frequency:cite:p:`buchsbaum:1960`., also called the bi-ion frequency or ion-ion hybrid frequency:cite:p:`vincena:2013`.
```
```suggestion
    Buchsbaum frequency :cite:p:`buchsbaum:1960`, also called the bi-ion frequency or ion-ion hybrid frequency :cite:p:`vincena:2013`.
```

svincena replied:
Also fixed later. Guess I made the PR to soon!

namurphy reviewed (commented):
Thank you for submitting this!  I have a few minor formatting suggestions.  Overall, this looks like a fantastic addition.

svincena commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,70 @@
+from plasmapy.particles import Particle
+
+
+def buchsbaum_frequency(B: Unit('T'), n1: u.m**-3, n2: u.m**-3, particle1: Particle, particle2: Particle, Z1=None,
```
This was fixed in a later commit.

svincena commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,70 @@
+        The Buchsbaum frequency of the plasma in units of radians per second.
+
+        Notes
+        -----
```
Also fixed later.  Guess I made the PR to soon!

svincena committed:
Update hack/Buchsbaum_frequency.py per Nick

Co-authored-by: Nick Murphy <namurphy@cfa.harvard.edu>

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,95 @@
+import numpy as np
+import astropy.units as u
```
```suggestion
__all__ = ["buchsbaum_frequency"]
__aliases__ = ["omega_bb_", "omega_ii_", "omega_bi_"]

import numpy as np
import astropy.units as u
```

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,95 @@
+from plasmapy.utils.decorators import (
+    angular_freq_to_hz,
+    validate_quantities,
+)
```
```suggestion
)

__all__ += __aliases__
```

svincena replied:
Thanks!

rocco8773 reviewed (changes requested):
Define `__all__` and `__aliases__` in preparation for documentation.

svincena committed:
Add __all__ and __aliases__

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
__all__ = ["buchsbaum_frequency"]
__aliases__ = ["o_bb_", "o_ii_", "o_bi_"]
per Erik

rocco8773 reviewed (changes requested):
Now you need to add a `hack.Buchsbaum_frequency.rst` file to `/docs/api_static/` with the following content...

```rst
:orphan:

`hack.Buchsbaum_frequency`
==========================

.. currentmodule:: hack.Buchsbaum_frequency

.. automodapi:: hack.Buchsbaum_frequency

```

svincena committed:
Add docs/api_static/hack.Buchsbaum_frequency.rst
per Erik

svincena reviewed (commented):
Made changes per Erik and Nick

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,98 @@
+
+    return np.sqrt((omega_p1**2 * omega_c2**2 + omega_p2**2 * omega_c1**2) / ( omega_p1**2 + omega_p2**2))
+
+omega_bb_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
+
+omega_ii_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
+
+omega_bi_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
```
```suggestion
o_bb_ = buchsbaum_frequency
"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""

o_ii_ = buchsbaum_frequency
"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""

o_bi_ = buchsbaum_frequency
"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
```

The miss-match naming to `__aliases__` is what's causing the doc build to currently fail.

svincena replied:
done. thanks.

svincena committed:
Changed omega_bb_, etc to o_bb_

svincena committed:
Changed o_bb_, etc to w_bb_, etc
I prefer this notation. Erik preferred 'o' for 'Omega'
but 'w' is used for lower case omega, which applies to this function as well.

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,98 @@
+                        Z1=None, Z2=None) -> u.rad / u.s:
+
+    r"""
+    Calculate the Buchsbaum frequency in units of radians per second.
+    **Aliases:** `omega_bb_`, `omega_ii_`, `omega_bi_`
+    Parameters
```
```suggestion
    Calculate the Buchsbaum frequency in units of radians per second.
    
    **Aliases:** `o_bb_`, `o_ii_`, `o_bi_`
    
    Parameters
```

We need spaces here so `sphinx` can properly parse, and the aliases need to be updated.

svincena committed:
Added some space per Erik & updated **Aliases** in docstring.

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,100 @@
+    omega_BB : `~astropy.units.Quantity`
+        The Buchsbaum frequency of the plasma in units of radians per second.
+        Setting keyword ``to_hz=True`` will apply the factor of :math:`1/2π`
+        and yield a value in Hz.
+    Raises
```
```suggestion
        and yield a value in Hz.
    
    Raises
```

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,100 @@
+        ``particle`` is not of an appropriate type.
+    `ValueError`
+        If the magnetic field contains invalid values or particle cannot
+        be used to identify a particle or isotope.
+    Warns
```
```suggestion
        be used to identify a particle or isotope.

    Warns
```

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,100 @@
+    .. math::
+    \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2} + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
+
+#   These are the references, but do they need to go into a global bibtex-type document?
+#   [1] S. J. Buchsbaum, Phys. Fluids 3, 418 (1960); Resonance in a Plasma with Two Ion Species; https://doi.org/10.1063/1.1706052
+#   [2] S. T. Vincena, W. A. Farmer, J. E. Maggs, and G. J. Morales, Phys. Plasmas, 20, 012111 (2013);
+#    Investigation of an ion-ion hybrid Alfvén wave resonator https://doi.org/10.1063/1.4775777
```
move these outside the docstring for now

svincena replied:
will do

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Moved commented-out lines out of docstring into function

svincena committed:
added references to bibliography.bib

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,107 @@
+    angular_freq_to_hz,
+    validate_quantities,
+)
+__all__ = ["buchsbaum_frequency"]
+__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]
```
```suggestion
```

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,107 @@
+import numpy as np
```
```suggestion
__all__ = ["buchsbaum_frequency"]
__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]

import numpy as np
```

rocco8773 reviewed (changes requested):
In practice, the `__all__` dunder is supposed to be defined at the top of the module, just after the module docstring.

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,107 @@
+)
+__all__ = ["buchsbaum_frequency"]
+__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]
+__all__ += __aliases__
+
+@validate_quantities(
```
```suggestion
__all__ += __aliases__


@validate_quantities(
```

2 blank lines before a function or class definition

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,107 @@
+    }
+)
+@angular_freq_to_hz
+def buchsbaum_frequency(B: u.T, n1: u.m**-3, n2: u.m**-3, particle1: Particle, particle2: Particle,
+                        Z1=None, Z2=None) -> u.rad / u.s:
+
+    r"""
```
```suggestion
def buchsbaum_frequency(
    B: u.T,
    n1: u.m**-3,
    n2: u.m**-3,
    particle1: Particle,
    particle2: Particle,
    Z1=None,
    Z2=None,
) -> u.rad / u.s:
    r"""
```

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,107 @@
+    r"""
+    Calculate the Buchsbaum frequency in units of radians per second.
+
+   **Aliases:** `w_bb_`, `w_ii_`, `w_bi_`
```
```suggestion
    **Aliases:** `w_bb_`, `w_ii_`, `w_bi_`
```

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,107 @@
+    omega_BB : `~astropy.units.Quantity`
+        The Buchsbaum frequency of the plasma in units of radians per second.
+        Setting keyword ``to_hz=True`` will apply the factor of :math:`1/2π`
+        and yield a value in Hz.
+    
+    Raises
```
```suggestion
        and yield a value in Hz.

    Raises
```

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,115 @@
+    component of the cold-plasma dielectric coefficient
+    math:: \epsilon_{\perp} to vanish at an angular frequency referred to as the
+    Buchsbaum frequency:cite:p:`buchsbaum:1960`., also called the bi-ion frequency or ion-ion hybrid frequency:cite:p:`vincena:2013`.
+    This frequency can be defined as:
```
```suggestion
    In a magnetized plasma, the presence of two ion species allows the
    perpendicular component of the cold-plasma dielectric coefficient
    :math:`\epsilon_{\perp}` to vanish at an angular frequency referred
    to as the Buchsbaum frequency :cite:p:`buchsbaum:1960`, also called
    the bi-ion frequency or ion-ion hybrid frequency
    :cite:p:`vincena:2013`.  This frequency can be defined as:
```

svincena committed:
deleted commented lines regarding references because they're in the bibliography.bib file now.

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,115 @@
+    Buchsbaum frequency:cite:p:`buchsbaum:1960`., also called the bi-ion frequency or ion-ion hybrid frequency:cite:p:`vincena:2013`.
+    This frequency can be defined as:
+
+    .. math::
+    \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2} + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
```
```suggestion
    .. math::
        \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2}
            + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
```

svincena committed:
72 character line enforcement

Co-authored-by: Erik Everson <eteverson@gmail.com>

svincena committed:
Update line length in long equation

Co-authored-by: Erik Everson <eteverson@gmail.com>

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,110 @@
+__aliases__ = ["w_bb_", "w_ii_", "w_bi_"]
+
+import numpy as np
+import astropy.units as u
+from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
```
```suggestion
import astropy.units as u

from plasmapy.formulary.frequencies import gyrofrequency, plasma_frequency
```

svincena committed:
Update hack/Buchsbaum_frequency.py

Co-authored-by: Erik Everson <eteverson@gmail.com>

rocco8773 commented on hack/Buchsbaum_frequency.py:
```diff
@@ -0,0 +1,112 @@
+
+
+w_bb_ = buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.buchsbaum_frequency`."""
```
```suggestion
"""Alias to `~hack.Buchsbaum_frequency.buchsbaum_frequency`."""
```

svincena committed:
Changed alias references to """Alias to `~hack.Buchsbaum_frequency.buchsbaum_frequency`."""
